File: app/src/main/java/cm/aptoide/pt/autoupdate/AutoUpdateRepository.java
package cm.aptoide.pt.autoupdate;

import rx.Single;

public class AutoUpdateRepository {

  private final AutoUpdateService autoUpdateService;
  private AutoUpdateModel autoUpdateModel;

  public AutoUpdateRepository(AutoUpdateService autoUpdateService) {
    this.autoUpdateService = autoUpdateService;
  }

  public Single<AutoUpdateModel> loadFreshAutoUpdateModel() {
    return loadAndSaveAutoUpdateModel();
  }

  public Single<AutoUpdateModel> loadAutoUpdateModel() {
    if (autoUpdateModel != null) {
      return Single.just(autoUpdateModel);
    }
    return loadAndSaveAutoUpdateModel();
  }

  private Single<AutoUpdateModel> loadAndSaveAutoUpdateModel() {
    return autoUpdateService.loadAutoUpdateModel()
        .doOnSuccess(autoUpdateModel -> this.autoUpdateModel = autoUpdateModel);
  }
}

File: old_app/src/main/java/cm/aptoide/pt/autoupdate/AutoUpdateRepository.kt
package cm.aptoide.pt.autoupdate

import rx.Single

class AutoUpdateRepository(private val autoUpdateService: AutoUpdateService) {

  private lateinit var autoUpdateModel: AutoUpdateModel

  fun loadFreshAutoUpdateModel(): Single<AutoUpdateModel> {
    return loadAndSaveAutoUpdateModel()
  }

  fun loadAutoUpdateModel(): Single<AutoUpdateModel> {
    return if (::autoUpdateModel.isInitialized) {
      Single.just(autoUpdateModel)
    } else
      loadAndSaveAutoUpdateModel()

  }

  private fun loadAndSaveAutoUpdateModel(): Single<AutoUpdateModel> {
    return autoUpdateService.loadAutoUpdateModel()
        .doOnSuccess { autoUpdateModel = it }
  }
}